### DISCUSSION TITLE
How does the auto-completion feature work?

### DISCUSSION THREAD

**Discussion by airvzxf** [shuru-project/shuru#21] [Q&A] [answered]
How does the auto-completion feature work?

I don't quite understand how auto-completion works; from my experience, I guess that I run the command _“shuru --completions bash”_ and the output has to be saved in the _“.bashrc”_ or _“.bash_profile”_ file. Is this correct?

If so, I wonder if it would be better if there was another option to automatically add this Bash function to the files mentioned in the paragraph above. This is what JetBrains does with PyCharm and others: it checks if the function exists, and if not, it adds it automatically.

Another option is creating an alias which sets once the command _“shuru”_ runs. This is not invasive, adding code in the bash files.

**Comment by harshdoesdev** [comment, marked as answer]
Yes, you’re correct that running `shuru --completions <SHELL>` generates the auto-completion script, and that output should be saved somewhere your shell loads, like `~/.bashrc`, `~/.bash_profile`, or a dedicated completion file. For example:

```bash
shuru --completions bash > ~/.bash_completion
```

We don’t automatically install the script because users often have different setups and permissions. Manual installation offers several benefits:

- **Flexibility**: Different users and systems have different setups. By printing the script, `shuru` allows users to decide where and how they want to install it, ensuring it fits into their specific environment. For example, users might want to install the completion script to a global location (e.g., `/etc/bash_completion.d/`) or a user-specific location (e.g., `~/.bash_completion`), depending on their needs and permissions.
- **Avoids Permission Issues**: Manual setup helps prevent problems related to permissions when modifying system files.
- **Preserves Custom Configurations**: It ensures existing shell configurations remain intact, avoiding unintended overrides.

That said, I find the idea of automatically adding the completion script, similar to JetBrains tools, interesting and will consider adding an option for easier setup in the future.

**Comment by airvzxf** [comment]
Thanks for the clarification and for sharing your perspective. I strongly agree. Maybe if we add this information in the documentation (manuals) or in the README, it will be enough, and then the user will be able to read it and execute it if he decides to implement it.
